pika/connection.py

```diff
@@ -1997,97 +2068,128 @@ def _on_data_available(self, data_in):
-                         'accessing a virtual host')
-            connection_error = exceptions.ProbableAccessDeniedError(
-                reason_code,
-                reason_text)
-        elif self.connection_state not in [self.CONNECTION_OPEN,
-                                           self.CONNECTION_CLOSED,
-                                           self.CONNECTION_CLOSING]:
-            LOGGER.warning('Unexpected connection state on disconnect: %i',
-                           self.connection_state)
+        if not self._opened and isinstance(self._error,
+                                           exceptions.StreamLostError):
+            # Heuristically deduce error based on connection state
+            if self.connection_state == self.CONNECTION_PROTOCOL:
+                LOGGER.error('Probably incompatible Protocol Versions')
+                self._error = exceptions.IncompatibleProtocolError(
+                    repr(self._error))
+            elif self.connection_state == self.CONNECTION_START:
+                LOGGER.error('Connection closed while authenticating indicating a '
+                             'probable authentication error')
+                self._error = exceptions.ProbableAuthenticationError(
+                    repr(self._error))
+            elif self.connection_state == self.CONNECTION_TUNE:
+                LOGGER.error('Connection closed while tuning the connection '
+                             'indicating a probable permission error when '
+                             'accessing a virtual host')
+                self._error = exceptions.ProbableAccessDeniedError(
+                    repr(self._error))
+            elif self.connection_state not in [self.CONNECTION_OPEN,
+                                               self.CONNECTION_CLOSED,
+                                               self.CONNECTION_CLOSING]:
+                LOGGER.warning('Unexpected connection state on disconnect: %i',
+                               self.connection_state)
 
         # Transition to closed state
         self._set_connection_state(self.CONNECTION_CLOSED)
 
-        # Inform our channel proxies
+        # Inform our channel proxies, if any are still around
         for channel in dictkeys(self._channels):
             if channel not in self._channels:
                 continue
             # pylint: disable=W0212
-            self._channels[channel]._on_close_meta(reason_code, reason_text)
+            self._channels[channel]._on_close_meta(
+                pika.channel.ClientChannelErrors.CONNECTION_CLOSED,
+                repr(self._error))
 
         # Inform interested parties
-        if connection_error is not None:
-            LOGGER.error('Connection setup failed due to %r', connection_error)
+        if not self._opened:
+            LOGGER.info('Connection setup terminated due to %r', self._error)
             self.callbacks.process(0,
```

Comment:
Orthogonally, only do one or the other callback depending on state, but not both.